### PR TITLE
feat(kube-api-rewriter): add debug server

### PR DESCRIPTION
**Pull request by yaroslavborbat**
## Description
- Add pprof server. Enable with PPROF_BIND_ADDRESS.
- Enable pprof when logLevel=debug.
- Add metrics server with builtin Go and process metrics.
- Add custom metrics to track rewriter proxy internals.

## Why do we need it, and what problem does it solve?

We want to diagnose problems with kube-api-rewriter container.

## What is the expected result?

- pprof can be enabled with logLevel=debug in mc/virtualization.
- Prometheus scrapes kube_api_rewriter_* metrics.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: images/kube-api-proxy/pkg/debug/server/server.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+	}
+	if s.pprofListener != nil {
+		s.addPprofServer()
+	}
+	s.runnableGroup.Start()
+}
+
+func (s *Server) Stop() {
+	s.runnableGroup.Start()
+}
```
Comment (diafour) [inline images/kube-api-proxy/pkg/debug/server/server.go:61]: Start inside Stop =)

File: templates/_kube_api_rewriter.tpl [hunk 2]
```diff
@@ -27,6 +31,12 @@ spec:
         configMap:
           defaultMode: 0644
           name: kube-api-proxy-kubeconfig
+{{- if eq .Values.virtualization.logLevel "debug" }}
```
Comment (diafour) [inline templates/_kube_api_rewriter.tpl:34]:
```suggestion
{{- if eq $ctx.Values.virtualization.logLevel "debug" }}
```

File: images/kube-api-proxy/pkg/rewriter/list.go [hunk 3]
```diff
@@ -52,30 +52,51 @@ var SkipItem = errors.New("remove item from the result")
...
 		return obj, nil
 	}
-	rwrItems := []byte(`[]`)
+
+	var rwrItems bytes.Buffer
+	rwrItems.Grow(len(obj))
+	// Start array
+	rwrItems.WriteString(`[`)
+
+	//rwrItems := []byte(`[]`)
```
Comment (Isteb4k) [inline images/kube-api-proxy/pkg/rewriter/list.go:72]: Delete comment

File: images/kube-api-proxy/pkg/rewriter/discovery.go [hunk 4]
```diff
@@ -390,35 +405,42 @@ func RestoreAggregatedGroupDiscovery(rules *RewriteRules, obj []byte) ([][]byte,
 		restoredGroupObj := []byte(fmt.Sprintf(`{"metadata":{"name":"%s", "creationTimestamp":null}}`, groupName))
 
 		// Construct an array of APIVersionDiscovery objects.
-		restoredVersions := []byte(`[]`)
+		//restoredVersions := []byte(`[]`)
```
Comment (Isteb4k) [inline images/kube-api-proxy/pkg/rewriter/discovery.go:408]: Old comment, looks like you can delete it

File: images/kube-api-proxy/pkg/proxy/stream_handler.go [hunk 5]
```diff
@@ -151,26 +189,27 @@ func (s *StreamHandler) proxy() {
...
 // Source
 // k8s.io/client-go@v0.26.1/rest/request.go:765 newStreamWatcher
 // k8s.io/apimachinery@v0.26.1/pkg/runtime/negotiate.go:70 StreamDecoder
-func (s *StreamHandler) createWatchDecoder(contentType string) (streaming.Decoder, error) {
+func createWatchDecoder(r io.Reader, contentType string) (streaming.Decoder, error) {
 	mediaType, _, err := mime.ParseMediaType(contentType)
 	if err != nil {
-		s.log.Error("Unexpected media type from the server: %q: %v", contentType, err)
+		return nil, fmt.Errorf("unexpected media type from the server: %q: %w", contentType, err)
+		//s.log.Error("Unexpected media type from the server: %q: %v", contentType, err)
```
Comment (Isteb4k) [inline images/kube-api-proxy/pkg/proxy/stream_handler.go:212]: Delete comment

File: images/kube-api-proxy/pkg/proxy/handler.go [hunk 6]
```diff
@@ -137,29 +157,45 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+		req.Body = BytesCounterReaderWrap(bytes.NewBuffer(origRequestBytes))
 	}
 
 	// Step 3. Send request to the target.
 	resp, err := h.TargetClient.Do(req)
 	if err != nil {
 		logger.Error("Error passing request to the target", logutil.SlogErr(err))
 		http.Error(w, "Error passing request to the target", http.StatusInternalServerError)
+		metrics.TargetResponseError()
 		// TODO return apimachinery NewInternalError
```
Comment (Isteb4k) [inline images/kube-api-proxy/pkg/proxy/handler.go:184]: Do you have task for this TODO?

File: images/kube-api-proxy/pkg/proxy/handler.go [hunk 7]
```diff
@@ -137,29 +157,45 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	if err != nil {
 		logger.Error(fmt.Sprintf("Error transforming request: %s", req.URL.String()), logutil.SlogErr(err))
 		http.Error(w, "can't rewrite request", http.StatusBadRequest)
+		// TODO add another more specific abstraction layer on top of MetricsProvider: h.metrics.incomingRewriteErrorInc()
```
Comment (Isteb4k) [inline images/kube-api-proxy/pkg/proxy/handler.go:160]: Create task for this TODO

**Review by Isteb4k — CHANGES_REQUESTED** [review]

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/kube-api-proxy/pkg/debug/runnablegroup/runnable.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+func (r *RunnableGroup) run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	wg := sync.WaitGroup{}
+	var mu sync.Mutex
+	var retErr error
+	for _, runnable := range r.runnable {
+		wg.Add(1)
+		runnable := runnable
```
[the remark this answers is not in the document]
Comment (diafour) [reply images/kube-api-proxy/pkg/debug/runnablegroup/runnable.go:60]: @yaroslavborbat fyi =)

File: images/kube-api-proxy/pkg/proxy/handler.go (hunk 7 above)
Comment (diafour) [reply images/kube-api-proxy/pkg/proxy/handler.go:160]: This TODO was implemented in ProxyMetrics in metrics.go

File: images/kube-api-proxy/pkg/proxy/stream_handler.go (hunk 5 above)
Comment (diafour) [reply images/kube-api-proxy/pkg/proxy/stream_handler.go:212]: Deleted

File: images/kube-api-proxy/pkg/rewriter/list.go (hunk 3 above)
Comment (diafour) [reply images/kube-api-proxy/pkg/rewriter/list.go:72]: Done

File: images/kube-api-proxy/pkg/rewriter/discovery.go (hunk 4 above)
Comment (diafour) [reply images/kube-api-proxy/pkg/rewriter/discovery.go:408]: Done

File: images/kube-api-proxy/pkg/proxy/handler.go (hunk 6 above)
Comment (diafour) [reply images/kube-api-proxy/pkg/proxy/handler.go:184]: Add card/39968591.

**Pull request merged**